cheroot/connections.py

```diff
@@ -84,9 +100,15 @@ def put(self, conn):
         if conn.rfile.has_data():
             self._readable_conns.append(conn)
         else:
-            self._selector.register(
-                conn.socket.fileno(), selectors.EVENT_READ, data=conn,
-            )
+            self._put_q.append(conn)
+            with self._ctrl_lock:
+                self._ctrl_tx.send(self._CTRL_MSG_PUT)
+
+    def _get_selector_conns(self):
+        # helper to retrieve conns registered with the selector
+        for _, (_, sock_fd, _, conn) in self._selector.get_map().items():
```

Comment:
the strategy in this PR is to ensure that only a single thread is ever touching the selector, so from a thread safety perspective it should not be necessary. do you have another safety case in mind?

Reply:
Because this is in a separate method, the callers would need to account for the implementation details. Hence this is unsafe and will probably introduce future bugs if somebody will attempt to refactor the calling code (turn two loops into one) w/o having the context.

Reply:
given that this is marked as a 'private' helper, i'd just as soon indicate in the docstring that this is returning references and callers should not store them, etc. if you feel strongly, it should be ok to copy them, but this is mainly intended just to provide a filter on the iteration of the map items.

Reply:
I understand this, I'm mostly thinking about preventing future regressions caused by the human factor.

Reply:
@webknjaz about the idea of copying the`get_map` return value, the internal map is not a dictionary, but rather is generated by making `__getitem__` calls, there is no `copy()` method (and the copy module will basically iterate over the content). The class of the object that is returned as a map is https://github.com/python/cpython/blob/master/Lib/selectors.py#L61-L78 which if you see the implementation of the `Mapping` abstract class, you'll see that the `get_map().items()` will end up in https://github.com/python/cpython/blob/master/Lib/_collections_abc.py#L764, and it works by iterating the keys and making the `__getitem__` method call. 

The alternative is to use the private dictionaries of that class, but I think that it would be a very bad idea to break that _interface contract_, that's why I preferred to manage an external set of key instead of having the overhead of iterating all the keys to make the copy (register/unregister returns the key) and the additional method calling, plus a lock to make sure we have a copy that can be iterated over in a threaded context.

Reply:
i agree that if the additional locking & copying are acceptable, the approach in https://github.com/cherrypy/cheroot/pull/318 from @cyraxjoe makes sense to me. this PR is predicated on the assumption that those are undesirable.

given the request to copy these to account for developer error in the future, it sounds like that might be preferable? either way is ok by me.